compiler/rustc_query_system/src/dep_graph/graph.rs

```diff
@@ -1367,6 +1345,36 @@ impl DepNodeColorMap {
         if value <= DepNodeIndex::MAX_AS_U32 { Some(DepNodeIndex::from_u32(value)) } else { None }
     }
 
+    /// This tries to atomically mark a node green and assign `index` as the new
+    /// index.
+    #[inline]
+    pub(super) fn try_mark_green(
+        &self,
+        prev_index: SerializedDepNodeIndex,
+        index: DepNodeIndex,
+    ) -> Result<(), DepNodeIndex> {
+        let value = &self.values[prev_index];
+        if self.sync {
+            match value.compare_exchange(
+                COMPRESSED_NONE,
+                index.as_u32(),
+                Ordering::Relaxed,
+                Ordering::Relaxed,
+            ) {
+                Ok(_) => Ok(()),
+                Err(v) => Err(DepNodeIndex::from_u32(v)),
+            }
+        } else {
```

Comment:
is it actually a meaningful improvement to not use compare_exchange in both code paths? It seems like in the single threaded case the compare exchange should never actually have to communicate across cores and thus be not more expensive than a load + store

Reply:
It roughly 1/3 of the cost based on the benchmarks I did for the branch for `Lock`, which is just 2x `compare_exchange` in the uncongested case.